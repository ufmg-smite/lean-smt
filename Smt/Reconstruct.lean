/-
Copyright (c) 2021-2023 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import cvc5
import Qq

import Smt.Attribute

namespace Smt

open Lean
open Attribute

structure Reconstruct.state where
  termCache : HashMap cvc5.Term Expr
  proofCache : HashMap cvc5.Proof Expr
  count : Nat
  currAssums : Array Expr
  skippedGoals : Array MVarId

abbrev ReconstructM := StateT Reconstruct.state MetaM

abbrev SortReconstructor := cvc5.Sort → ReconstructM (Option Expr)

abbrev TermReconstructor := cvc5.Term → ReconstructM (Option Expr)

abbrev ProofReconstructor := cvc5.Proof → ReconstructM (Option Expr)

namespace Reconstruct

private unsafe def getReconstructorsUnsafe (n : Name) (rcons : Type) : MetaM (List (rcons × Name)) := do
  let env ← getEnv
  let names := ((smtExt.getState env).findD n {}).toList
  let mut reconstructors := []
  for name in names do
    let fn ← IO.ofExcept <| Id.run <| ExceptT.run <|
      env.evalConst rcons Options.empty name
    reconstructors := (fn, name) :: reconstructors
  return reconstructors

@[implemented_by getReconstructorsUnsafe]
opaque getReconstructors (n : Name) (rcons : Type) : MetaM (List (rcons × Name))

def reconstructSort (s : cvc5.Sort) : ReconstructM Expr := do
  let rs ← getReconstructors ``SortReconstructor SortReconstructor
  go rs s
where
  go (rs : List (SortReconstructor × Name)) (s : cvc5.Sort) : ReconstructM Expr := do
    for (r, n) in rs do
      if let some e ← r s then
        trace[smt.reconstruct.sort] "{s} =({n})=> {e}"
        return e
    throwError "Failed to reconstruct sort {s} with kind {s.getKind}"

def traceReconstructTerm (t : cvc5.Term) (r : Except Exception Expr) : ReconstructM MessageData :=
  return m!"{t} ↦ " ++ match r with
    | .ok e    => m!"{e}"
    | .error _ => m!"{bombEmoji}"

def withNewTermCache (k : ReconstructM α) : ReconstructM α := do
  let state ← get
  let r ← k
  set { ← get with termCache := state.termCache }
  return r

def reconstructTerm : cvc5.Term → ReconstructM Expr := withTermCache fun t => do
  withTraceNode `smt.reconstruct.term (traceReconstructTerm t) do
    let rs ← getReconstructors ``TermReconstructor TermReconstructor
    go rs t
where
  withTermCache (r : cvc5.Term → ReconstructM Expr) (t : cvc5.Term) : ReconstructM Expr := do
    match (← get).termCache.find? t with
    -- TODO: cvc5's global bound variables mess up the cache. Find a better fix.
    | some e => return e
    | none   => reconstruct r t
  reconstruct r t := do
    let e ← r t
    modify fun state => { state with termCache := state.termCache.insert t e }
    return e
  go (rs : List (TermReconstructor × Name)) (t : cvc5.Term) : ReconstructM Expr := do
    for (r, n) in rs do
      if let some e ← r t then
        trace[smt.reconstruct.term] "{t} =({n})=> {e}"
        return e
    throwError "Failed to reconstruct term {t} with kind {t.getKind}"

def withNewProofCache (k : ReconstructM α) : ReconstructM α := do
  let state ← get
  let r ← k
  set { ← get with proofCache := state.proofCache }
  return r

def withProofCache (r : cvc5.Proof → ReconstructM Expr) (pf : cvc5.Proof) : ReconstructM Expr := do
  match (← get).proofCache.find? pf with
  | some e => return e
  | none   => reconstruct r pf
where
  reconstruct r pf := do
    let e ← r pf
    modify fun state => { state with proofCache := state.proofCache.insert pf e }
    return e

def incCount : ReconstructM Nat :=
  modifyGet fun state => (state.count, { state with count := state.count + 1 })

def withAssums (as : Array Expr) (k : ReconstructM α) : ReconstructM α := do
  modify fun state => { state with currAssums := state.currAssums ++ as }
  let r ← k
  modify fun state => { state with currAssums := state.currAssums.shrink (state.currAssums.size - as.size) }
  return r

def findAssumWithType? (t : Expr) : ReconstructM (Option Expr) := do
  for a in (← get).currAssums.reverse do
    let t' ← a.fvarId!.getType
    if t' == t then
      return some a
  return none

def skipStep (mv : MVarId) : ReconstructM Unit := mv.withContext do
  let state ← get
  let t ← Meta.mkForallFVars state.currAssums (← mv.getType)
  let ctx := state.currAssums.foldr (fun e ctx => ctx.erase e.fvarId!) (← getLCtx)
  let mv' ← Meta.withLCtx ctx (← Meta.getLocalInstances) (Meta.mkFreshExprMVar t)
  let e := mkAppN mv' state.currAssums
  set { state with skippedGoals := state.skippedGoals.push mv'.mvarId! }
  mv.assign e

def addThm (type : Expr) (val : Expr) : ReconstructM Expr := do
  let name := Name.num `s (← incCount)
  let mv ← Meta.mkFreshExprMVar type .natural name
  mv.mvarId!.assign val
  trace[smt.reconstruct.proof] "have {name} : {type} := {mv}"
  return mv

def addTac (type : Expr) (tac : MVarId → MetaM Unit) : ReconstructM Expr := do
  let name := Name.num `s (← incCount)
  let mv ← Meta.mkFreshExprMVar type .natural name
  tac mv.mvarId!
  trace[smt.reconstruct.proof] "have {name} : {type} := {mv}"
  return mv

def addTrust (type : Expr) (pf : cvc5.Proof) : ReconstructM Expr := do
  let name := Name.num `s (← incCount)
  let mv ← Meta.mkFreshExprMVar type .natural name
  skipStep mv.mvarId!
  trace[smt.reconstruct.proof] m!"have {name} : {type} := sorry"
  trace[smt.reconstruct.proof]
    m!"rule : {pf.getRule}\npremises : {pf.getChildren.map (·.getResult)}\nargs : {pf.getArguments}\nconclusion : {pf.getResult}"
  return mv

partial def reconstructProof : cvc5.Proof → ReconstructM Expr := withProofCache fun pf => do
  let rs ← getReconstructors ``ProofReconstructor ProofReconstructor
  go rs pf
where
  go (rs : List (ProofReconstructor × Name)) (pf : cvc5.Proof) : ReconstructM Expr := do
    for (r, _) in rs do
      if let some e ← r pf then
        return e
    _ ← pf.getChildren.mapM reconstructProof
    let type ← reconstructTerm pf.getResult
    addTrust type pf

end Reconstruct

def traceReconstructProof (r : Except Exception (Expr × Expr × List MVarId)) : MetaM MessageData :=
  return match r with
  | .ok _ => m!"{checkEmoji}"
  | _     => m!"{bombEmoji}"

open Qq in
partial def reconstructProof (pf : cvc5.Proof) : MetaM (Expr × Expr × List MVarId) := do
  withTraceNode `smt.reconstruct.proof traceReconstructProof do
  let Prod.mk (p : Q(Prop)) state ← (Reconstruct.reconstructTerm (pf.getResult)).run ⟨{}, {}, 0, #[], #[]⟩
  let Prod.mk (h : Q(True → $p)) (.mk _ _ _ _ mvs) ← (Reconstruct.reconstructProof pf).run state
  return (p, q($h trivial), mvs.toList)

open cvc5 in
def traceSolve (r : Except Exception (Except SolverError Proof)) : MetaM MessageData :=
  return match r with
  | .ok (.ok _) => m!"{checkEmoji}"
  | _           => m!"{bombEmoji}"

open cvc5 in
def solve (query : String) (timeout : Option Nat) : MetaM (Except Error cvc5.Proof) :=
  profileitM Exception "simp" {} do
  withTraceNode `smt.solve traceSolve do Solver.run (← TermManager.new) do
    if let .some timeout := timeout then
      Solver.setOption "tlimit" (toString (1000*timeout))
    Solver.setOption "dag-thresh" "0"
    Solver.setOption "simplification" "none"
    Solver.setOption "enum-inst" "true"
    Solver.setOption "produce-models" "true"
    Solver.setOption "produce-proofs" "true"
    Solver.setOption "proof-elim-subtypes" "true"
    Solver.setOption "proof-granularity" "dsl-rewrite"
    Solver.parse query
    let r ← Solver.checkSat
    trace[smt.solve] m!"result: {r}"
    if r.isUnsat then
      let ps ← Solver.getProof
      if h : 0 < ps.size then
        return ps[0]
    throw (self := instMonadExcept _ _) (Error.user_error "something went wrong")

syntax (name := reconstruct) "reconstruct" str : tactic

open Lean.Elab Tactic in
@[tactic reconstruct] def evalReconstruct : Tactic := fun stx =>
  withMainContext do
    let some query := stx[1].isStrLit? | throwError "expected string"
    let r ← solve query none
    match r with
      | .error e => logInfo (repr e)
      | .ok pf =>
        let (p, hp, mvs) ← reconstructProof pf
        let mv ← Tactic.getMainGoal
        let mv ← mv.assert (Name.num `s 0) p hp
        let (_, mv) ← mv.intro1
        replaceMainGoal (mv :: mvs)

end Smt
