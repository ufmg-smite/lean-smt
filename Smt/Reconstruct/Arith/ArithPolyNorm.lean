/-
Copyright (c) 2021-2024 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Mathlib.Data.Real.Basic

namespace Smt.Arith

theorem lt_eq_sub_lt_zero [LinearOrderedRing α] {a b : α} : (a < b) = (a - b < 0) := by
  simp only [sub_neg]

theorem le_eq_sub_le_zero [LinearOrderedRing α] {a b : α} : (a ≤ b) = (a - b ≤ 0) := by
  simp only [tsub_le_iff_right, zero_add]

theorem eq_eq_sub_eq_zero [LinearOrderedRing α] {a b : α} : (a = b) = (a - b = 0) := by
  simp only [sub_eq_zero]

theorem ge_eq_sub_ge_zero [LinearOrderedRing α] {a b : α} : (a ≥ b) = (a - b ≥ 0) := by
  simp only [ge_iff_le, sub_nonneg]

theorem gt_eq_sub_gt_zero [LinearOrderedRing α] {a b : α} : (a > b) = (a - b > 0) := by
  simp only [gt_iff_lt, sub_pos]

theorem lt_of_sub_eq [LinearOrderedRing α] {c₁ c₂ a₁ a₂ b₁ b₂ : α} (hc₁ : c₁ > 0) (hc₂ : c₂ > 0) (h : c₁ * (a₁ - a₂) = c₂ * (b₁ - b₂)) : (a₁ < a₂) = (b₁ < b₂) := by
  have {c x y : α} (hc : c > 0) : (c * (x - y) < 0) = (x - y < 0) := by
    rw (config := { occs := .pos [1] }) [← mul_zero c, mul_lt_mul_left hc]
  rw [lt_eq_sub_lt_zero, @lt_eq_sub_lt_zero _ _ b₁, ← this hc₁, ← this hc₂, h]

theorem le_of_sub_eq [LinearOrderedRing α] {c₁ c₂ a₁ a₂ b₁ b₂ : α} (hc₁ : c₁ > 0) (hc₂ : c₂ > 0) (h : c₁ * (a₁ - a₂) = c₂ * (b₁ - b₂)) : (a₁ ≤ a₂) = (b₁ ≤ b₂) := by
  have {c x y : α} (hc : c > 0) : (c * (x - y) ≤ 0) = (x - y ≤ 0) := by
    rw (config := { occs := .pos [1] }) [← mul_zero c, mul_le_mul_left hc]
  rw [le_eq_sub_le_zero, @le_eq_sub_le_zero _ _ b₁, ← this hc₁, ← this hc₂, h]

theorem eq_of_sub_eq [LinearOrderedRing α] {c₁ c₂ a₁ a₂ b₁ b₂ : α} (hc₁ : c₁ > 0) (hc₂ : c₂ > 0) (h : c₁ * (a₁ - a₂) = c₂ * (b₁ - b₂)) : (a₁ = a₂) = (b₁ = b₂) := by
  have {c x y : α} (hc : c > 0) : (c * (x - y) = 0) = (x - y = 0) := by
    rw (config := { occs := .pos [1] }) [← mul_zero c, mul_right_inj' (ne_of_gt hc)]
  rw [@eq_eq_sub_eq_zero _ _ a₁, @eq_eq_sub_eq_zero _ _ b₁, ← this hc₁, ← this hc₂, h]

theorem ge_of_sub_eq [LinearOrderedRing α] {c₁ c₂ a₁ a₂ b₁ b₂ : α} (hc₁ : c₁ > 0) (hc₂ : c₂ > 0) (h : c₁ * (a₁ - a₂) = c₂ * (b₁ - b₂)) : (a₁ ≥ a₂) = (b₁ ≥ b₂) := by
  have {c x y : α} (hc : c > 0) : (c * (x - y) ≥ 0) = (x - y ≥ 0) := by
    rw (config := { occs := .pos [1] }) [← mul_zero c, ge_iff_le, mul_le_mul_left hc]
  rw [ge_eq_sub_ge_zero, @ge_eq_sub_ge_zero _ _ b₁, ← this hc₁, ← this hc₂, h]

theorem gt_of_sub_eq [LinearOrderedRing α] {c₁ c₂ a₁ a₂ b₁ b₂ : α} (hc₁ : c₁ > 0) (hc₂ : c₂ > 0) (h : c₁ * (a₁ - a₂) = c₂ * (b₁ - b₂)) : (a₁ > a₂) = (b₁ > b₂) := by
  have {c x y : α} (hc : c > 0) : (c * (x - y) > 0) = (x - y > 0) := by
    rw (config := { occs := .pos [1] }) [← mul_zero c, gt_iff_lt, mul_lt_mul_left hc]
  rw [gt_eq_sub_gt_zero, @gt_eq_sub_gt_zero _ _ b₁, ← this hc₁, ← this hc₂, h]

open Lean

open Lean Mathlib.Tactic.RingNF in
/-- Use `arithPolyNormCore` to rewrite the main goal. -/
def arithPolyNormCore (mv : MVarId) : MetaM (Option MVarId) := mv.withContext do
  let tgt ← instantiateMVars (← mv.getType)
  let s ← IO.mkRef {}
  let r ← M.run s {} <| rewrite tgt
  if r.expr.consumeMData.isConstOf ``True then
    mv.assign (← Meta.mkOfEqTrue (← r.getProof))
    return none
  else
    Meta.applySimpResultToTarget mv tgt r


def traceArithPolyNorm (r : Except Exception Unit) : MetaM MessageData :=
  return match r with
  | .ok _ => m!"{checkEmoji}"
  | _     => m!"{bombEmoji}"

/-- Use `arithPolyNorm` to rewrite the main goal. -/
def arithPolyNorm (mv : MVarId) : MetaM Unit := withTraceNode `smt.reconstruct.arithPolyNorm traceArithPolyNorm do
  mv.withContext do
  if let .some mv ← arithPolyNormCore mv then
    throwError "[arithPolyNorm]: could not prove {← mv.getType}"

def traceArithNormNum (r : Except Exception Unit) : MetaM MessageData :=
  return match r with
  | .ok _ => m!"{checkEmoji}"
  | _     => m!"{bombEmoji}"

open Mathlib.Meta.NormNum in
def normNum (mv : MVarId) : MetaM Unit := withTraceNode `smt.reconstruct.normNum traceArithNormNum do
  if let some (_, mv) ← normNumAt mv {} #[] true false then
    throwError "[norm_num]: could not prove {← mv.getType}"

open Qq in
partial def findConst (α : Q(Type)) (hα : Q(LinearOrderedRing $α)) (e : Q($α)) : MetaM Expr := do
  match e with
  | ~q($a * $b) => findConst α hα b
  | ~q($a + $b) => findConst α hα b
  | ~q($a - $b) => findConst α hα b
  | ~q(-$a)     => findConst α hα a
  | _           =>
    if e.hasFVar || e.hasLooseBVars then
      return q(1 : $α)
    else if e.getUsedConstants.contains ``Neg.neg then
      let e : Q(Real) := e
      match e with
      | ~q((-$a) / $b) => return q($a / $b)
      | _              => return e
    else
      return e

end Smt.Arith
