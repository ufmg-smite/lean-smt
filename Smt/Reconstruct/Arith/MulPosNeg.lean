/-
Copyright (c) 2021-2023 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Tomaz Gomes Mascarenhas
-/

-- implementation of:
-- https://cvc5.github.io/docs/cvc5-1.0.2/proofs/proof_rules.html#_CPPv4N4cvc58internal6PfRule14ARITH_MULT_POSE
-- and
-- https://cvc5.github.io/docs/cvc5-1.0.2/proofs/proof_rules.html#_CPPv4N4cvc58internal6PfRule14ARITH_MULT_NEGE

import Smt.Reconstruct.Arith.MulPosNeg.Lemmas
import Smt.Reconstruct.Arith.MulPosNeg.Tactic
