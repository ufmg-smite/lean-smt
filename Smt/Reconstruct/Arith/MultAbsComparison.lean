/-
Copyright (c) 2021-2025 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Tomaz Gomes Mascarenhas
-/

-- implementation of:
-- https://cvc5.github.io/docs/cvc5-1.2.1/api/cpp/enums/proofrule.html#_CPPv4N4cvc59ProofRule25ARITH_MULT_ABS_COMPARISONE

import Smt.Reconstruct.Arith.MultAbsComparison.Lemmas
