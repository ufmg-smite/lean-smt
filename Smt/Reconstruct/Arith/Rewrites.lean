/-
Copyright (c) 2021-2023 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Mathlib.Algebra.Order.Ring.Defs

namespace Smt.Reconstruct.Arith

-- https://github.com/cvc5/cvc5/blob/proof-new/src/theory/arith/rewrites

open Function

variable {α : Type} [h : LinearOrderedRing α]

variable {t ts x xs : α}

theorem arith_plus_zero : ts + 0 + ss = ts + ss :=
  (add_zero ts).symm ▸ rfl
theorem arith_mul_one : ts * 1 * ss = ts * ss :=
  (mul_one ts).symm ▸ rfl
theorem arith_mul_zero : ts * 0 * ss = 0 :=
  (mul_zero ts).symm ▸ (zero_mul ss).symm ▸ rfl

theorem arith_int_div_one {t : Int} : t / 1 = t :=
  Int.ediv_one t

theorem arith_neg_neg_one : -1 * (-1 * t) = t :=
  neg_mul _ t ▸ (one_mul t).symm ▸ neg_mul_neg _ t ▸ (one_mul t).symm ▸ rfl

-- Eliminations

theorem arith_elim_uminus : -t = -1 * t :=
  neg_eq_neg_one_mul t ▸ rfl
theorem arith_elim_minus : t - s = t + -1 * s :=
  neg_eq_neg_one_mul s ▸ sub_eq_add_neg _ s ▸ rfl
theorem arith_elim_gt : (t > s) = ¬(t ≤ s) :=
  propext not_le.symm
theorem arith_elim_lt : (t < s) = ¬(t ≥ s) :=
  propext not_le.symm
theorem arith_elim_leq : (t ≤ s) = (s ≥ t) :=
  propext ge_iff_le

theorem arith_leq_norm {t s : Int} : (t ≤ s) = ¬(t ≥ s + 1) :=
  propext ⟨(propext Int.not_le ▸ Int.lt_add_one_of_le ·),
           (Int.le_of_lt_add_one $ propext Int.not_le ▸ · )⟩

theorem arith_geq_tighten {t s : Int} : (¬(t ≥ s)) = (s ≥ t + 1) :=
  propext Int.not_le

theorem arith_geq_norm: (t ≥ s) = (t - s ≥ 0) :=
  propext ⟨sub_nonneg_of_le, le_of_sub_nonneg⟩

theorem arith_refl_leq : (t ≤ t) = True :=
  propext ⟨const _ trivial, const _ (le_refl t)⟩
theorem arith_refl_lt : (t < t) = False :=
  propext ⟨(lt_irrefl t), False.elim⟩
theorem arith_refl_geq : (t ≥ t) = True :=
  propext ⟨const _ trivial, const _ (le_refl t)⟩
theorem arith_refl_gt : (t > t) = False :=
  propext ⟨(lt_irrefl t), False.elim⟩

theorem arith_plus_flatten : xs + (w + ys) + zs = xs + w + ys + zs :=
  add_assoc xs w ys ▸ rfl

theorem arith_mult_flatten : xs * (w * ys) * zs = xs * w * ys * zs :=
  mul_assoc xs w ys ▸ rfl

theorem arith_mult_dist : x * (y + z + ws) = x * y + x * (z + ws) :=
  add_assoc y z ws ▸ mul_add x y (z + ws) ▸ rfl

theorem arith_plus_cancel1 : ts + x + ss + (-1 * x) + rs = ts + ss + rs :=
  neg_eq_neg_one_mul x ▸ add_assoc ts x ss ▸ add_assoc ts (x + ss) (-x) ▸
  add_comm x ss ▸ (add_neg_cancel_right ss x).symm ▸ rfl

theorem arith_plus_cancel2 : ts + (-1 * x) + ss + x + rs = ts + ss + rs :=
  neg_eq_neg_one_mul x ▸ add_assoc ts (-x) ss ▸ add_assoc ts (-x + ss) x ▸
  add_comm (-x) ss ▸ (neg_add_cancel_right ss x).symm ▸ rfl

end Smt.Reconstruct.Arith
