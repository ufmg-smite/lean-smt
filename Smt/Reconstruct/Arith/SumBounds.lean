/-
Copyright (c) 2021-2023 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Tomaz Gomes Mascarenhas
-/

import Smt.Reconstruct.Arith.SumBounds.Lemmas
import Smt.Reconstruct.Arith.SumBounds.Tactic
