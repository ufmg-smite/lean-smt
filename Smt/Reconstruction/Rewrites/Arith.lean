/-
Copyright (c) 2021-2023 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Std.Data.Int.Lemmas
import Smt.Reconstruction.Rewrites.Simp

namespace Smt.Reconstruction.Rewrites.Arith

scoped notation "const " c => fun _ => c

variable {xs w ys zs t x s r : Int}

-- Basic Rules

@[smt_simp] theorem arith_plus_zero : t + 0 + s = t + s :=
  t.add_zero.symm ▸ rfl

@[smt_simp] theorem arith_mul_one : t * 1 * s = t * s :=
  t.mul_one.symm ▸ rfl
@[smt_simp] theorem arith_mul_zero : t * 0 * s = 0 :=
  t.mul_zero.symm ▸ s.zero_mul.symm ▸ rfl

@[smt_simp] theorem arith_uminus_elim : -t = -1 * t :=
  t.neg_eq_neg_one_mul
@[smt_simp] theorem arith_neg_neg_one : -1 * (-1 * t) = t :=
  Int.mul_assoc (-1) (-1) t ▸ t.one_mul.symm ▸ rfl

-- Eliminations

@[smt_simp] theorem arith_elim_minus : t - s = t + -1 * s :=
  s.neg_eq_neg_one_mul ▸ rfl
@[smt_simp] theorem arith_elim_gt : (t > s) = ¬(t ≤ s) :=
  propext Int.not_le.symm
@[smt_simp] theorem arith_elim_lt : (t < s) = ¬(t ≥ s) :=
  propext Int.not_le.symm
@[smt_simp] theorem arith_elim_leq : (t ≤ s) = ¬(t ≥ s + 1) :=
  propext Int.not_lt.symm

@[smt_simp] theorem arith_geq_tighten : (¬(t ≥ s)) = (s ≥ t + 1) :=
  propext Int.not_le

@[smt_simp] theorem arith_geq_norm: (t ≥ s) = (t - s ≥ 0) :=
  propext ⟨Int.sub_nonneg_of_le, Int.le_of_sub_nonneg⟩

@[smt_simp] theorem arith_refl_leq : (t ≤ t) = True :=
  propext ⟨const trivial, const t.le_refl⟩
@[smt_simp] theorem arith_refl_lt : (t < t) = False :=
  propext ⟨t.lt_irrefl, False.elim⟩
@[smt_simp] theorem arith_refl_geq : (t ≥ t) = True :=
  propext ⟨const trivial, const t.le_refl⟩
@[smt_simp] theorem arith_refl_gt : (t > t) = False :=
  propext ⟨t.lt_irrefl, False.elim⟩

@[smt_simp] theorem plus_flatten : xs + (w + ys) + zs = xs + w + ys + zs :=
  Int.add_assoc xs w ys ▸ rfl

@[smt_simp] theorem mult_flatten : xs * (w * ys) * zs = xs * w * ys * zs :=
  Int.mul_assoc xs w ys ▸ rfl

@[smt_simp] theorem mult_dist : x * (y + z + w) = x * y + x * (z + w) :=
  Int.add_assoc y z w ▸ Int.mul_add x y (z + w) ▸ rfl

-- TEMPORARY

@[smt_simp] theorem arith_plus_cancel1 : t + x + s + (-1 * x) + r = t + s + r :=
  arith_uminus_elim ▸ Int.add_assoc t x s ▸ Int.add_assoc t (x + s) (-x) ▸
  Int.add_comm x s ▸ (Int.add_neg_cancel_right s x).symm ▸ rfl

@[smt_simp] theorem arith_plus_cancel2 : t + (-1 * x) + s + x + r = t + s + r :=
  arith_uminus_elim ▸ Int.add_assoc t (-x) s ▸ Int.add_assoc t (-x + s) x ▸
  Int.add_comm (-x) s ▸ (Int.neg_add_cancel_right s x).symm ▸ rfl

end Smt.Reconstruction.Rewrites.Arith
