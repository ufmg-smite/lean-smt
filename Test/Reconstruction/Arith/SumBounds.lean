import Smt.Reconstruct.Arith

open Smt.Reconstruct.Arith

example {a b c d e f : Nat} : a < d → b < e → c < f → a + (b + c) < d + (e + f) := by
  intros h₁ h₂ h₃
  sum_bounds [h₁, h₂, h₃]

example {a b c d e f w z : Int} :
  a ≤ d → b ≤ e → c ≤ f → w ≤ z → a + (b + (c + w)) ≤ d + (e + (f + z)) := by
    intros h₁ h₂ h₃ h₄
    sum_bounds [h₁, h₂, h₃, h₄]

example {a d : Int} {b c e f : Real} : a = d → b < e → c ≤ f → a + (b + c) < d + (e + f) := by
  intros h₁ h₂ h₃
  sum_bounds [h₁, h₂, h₃]

example {a b c d e f : Nat} : a < d → b < e → c < f → a + (b + c) < d + (e + f) := by
  intros h₁ h₂ h₃
  sum_bounds [h₁, h₂, h₃]
