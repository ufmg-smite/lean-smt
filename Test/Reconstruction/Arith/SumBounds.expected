Test/Reconstruction/Arith/SumBounds.lean:7:2: error: application type mismatch
  sumBounds₁ h₃
argument
  h₃
has type
  @LT.lt ℕ instLTNat c f : Prop
but is expected to have type
  @LT.lt ℕ Preorder.toLT c f : Prop
Test/Reconstruction/Arith/SumBounds.lean:12:4: error: type mismatch
  _uniq.452
has type
  w + c + b + a ≤ z + f + e + d : Prop
but is expected to have type
  a + (b + (c + w)) ≤ d + (e + (f + z)) : Prop
Test/Reconstruction/Arith/SumBounds.lean:16:2: error: type mismatch
  _uniq.1532
has type
  c + b + IntCast.intCast a < f + e + IntCast.intCast d : Prop
but is expected to have type
  ↑a + (b + c) < ↑d + (e + f) : Prop
Test/Reconstruction/Arith/SumBounds.lean:20:2: error: application type mismatch
  sumBounds₁ h₃
argument
  h₃
has type
  @LT.lt ℕ instLTNat c f : Prop
but is expected to have type
  @LT.lt ℕ Preorder.toLT c f : Prop
